### PR TITLE
API Allow controller discovery without underscores (PSR-2 compliance)

### LINKED ISSUE #6403: Allow controller discovery without underscore (PSR-2 compliance)
The `SiteTree->getControllerName()` (https://github.com/silverstripe/silverstripe-cms/blob/4.0.0-alpha3/code/Model/SiteTree.php#L2714-L2734) assumes an underscore in the class name, e.g. `Page_Controller`, which goes against the PSR-2 conventions (see [discussion](https://github.com/silverstripe/silverstripe-framework/pull/6379#issuecomment-267014124)).

Change the method to allow for non-underscored controller names, and deprecate the use of underscores (but keep it functional).

There's a separate PR for changing docs code examples accordingly: https://github.com/silverstripe/silverstripe-framework/issues/6376

### PR DESCRIPTION
**Pull request by robbieaverill**
* Update docs that reference `*_Controller` to `*Controller`.
* Reformat some docs for PSR-2 compat (relates to #6376). I didn't reformat them all, just a few of them. Still a fair bit of work needed for this issue.
* Update `Security` to pull `Page` and `PageController` from the global namespace instead of importing, in case the CMS is not present.
* Update `SSViewer` to resolve template names for `_Controller` and `Controller`. Use regex match to prevent classes like `SilverStripe\Control\Controller` from being matched by accident.
* Remove some legacy PHP code from the `SSViewerTest` class (e.g. `$self = $this` for closures).

---

* Resolves #6403.

## Blockers

- [ ] silverstripe/silverstripe-cms#1703

## Blocks

- [ ] silverstripe/silverstripe-installer#146

### CODE REVIEW

File: src/View/SSViewer.php [hunk 1]
```diff
@@ -253,9 +253,9 @@ public static function get_templates_by_class($classOrObject, $suffix = '', $bas
             $templates[] = $template;
             $templates[] = ['type' => 'Includes', $template];
 
-            // If the class is "Page_Controller", look for Page.ss
-            if (stripos($class, '_controller') !== false) {
-                $templates[] = str_ireplace('_controller', '', $class) . $suffix;
+            // If the class is "PageController" or "Page_Controller" (legacy), look for Page.ss
+            if (preg_match('/\w+_?Controller$/i', $class)) {
```
Comment (tractorcow) [inline src/View/SSViewer.php]: If you are using a regexp to match, use the same regexp to replace.

```php
if (preg_match('/^(?<name>\w+)(_?Controller)$/i', $class, $matches)) {
	$templates[] = $matches['name'] . $suffix;
}
```

Comment (robbieaverill) [inline src/View/SSViewer.php]: Makes sense @tractorcow - I've pushed an updated commit

**Comment by tractorcow** [comment]
I think I made one mistake... `\w` isn't going to capture namespace separators in the FQN. :) You might need to swap that for something more optimistic. Maybe just `.+`?

**Comment by sminnee** [comment]
The upgrader meta-data file should have an instruction to rewrite `Page_Controller` to `PageController`.

**Comment by robbieaverill** [comment]
@sminnee I've added the upgrade map, rebased and pushed

File: src/View/SSViewer.php [hunk 2]
```diff
@@ -253,9 +253,9 @@ public static function get_templates_by_class($classOrObject, $suffix = '', $bas
             $templates[] = $template;
             $templates[] = ['type' => 'Includes', $template];
 
-            // If the class is "Page_Controller", look for Page.ss
-            if (stripos($class, '_controller') !== false) {
-                $templates[] = str_ireplace('_controller', '', $class) . $suffix;
+            // If the class is "PageController" (PSR-2 compatibility) or "Page_Controller" (legacy), look for Page.ss
+            if (preg_match('/^(?<name>.+[^\\])_?Controller$/i', $class, $matches)) {
```
Comment (tractorcow) [inline src/View/SSViewer.php]: Sorry, you'll need to quad-escape this `[^\\\\]`. Which collapses to `[^\\]` when evaluated by PHP, and `[^\]` when evaluated by regexp :D

**Comment by tractorcow** [comment]
Cool, will merge if these pass. :)

**Comment by robbieaverill** [comment]
@tractorcow I'd assumed it requires silverstripe/silverstripe-cms#1703 first, but we'll see :)

**Comment by chillu** [comment]
@robbieaverill Do we need to mention this in the upgrading guide?

**Comment by robbieaverill** [comment]
@chillu yes, I'd say so, since it's a deprecating change. I'll be looking at those docs at some stage this week, but will update my Google doc (ref: #6441) for now

**Pull request merged**